core/state/snapshot/generate.go

```diff
@@ -159,169 +202,493 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
+		if len(keys) == max {
+			diskMore = iter.Next()
+			break
+		}
+	}
+	// Update metrics for database iteration and merkle proving
+	if kind == "storage" {
+		storageSnapRead += time.Since(start)
+		snapStorageSnapReadTimer.Update(int64(storageSnapRead))
+	} else {
+		accountSnapRead += time.Since(start)
+		snapAccountSnapReadTimer.Update(int64(accountSnapRead))
+	}
+	defer func(start time.Time) {
+		if kind == "storage" {
+			storageProving += time.Since(start)
+			snapStorageProveTimer.Update(int64(storageProving))
+		} else {
+			accountProving += time.Since(start)
+			snapAccountProveTimer.Update(int64(accountProving))
+		}
+	}(time.Now())
+
+	// The snap state is exhausted, pass the entire key/val set for verification
+	if origin == nil && !diskMore {
+		stackTr := trie.NewStackTrie(nil)
+		for i, key := range keys {
+			stackTr.TryUpdate(key, common.CopyBytes(vals[i]))
+		}
+		if gotRoot := stackTr.Hash(); gotRoot != root {
+			return &proofResult{keys: keys, vals: vals, diskMore: false, trieMore: false, proofErr: errors.New("wrong root")}, nil
+		}
+		return &proofResult{keys: keys, vals: vals, diskMore: false, trieMore: false, proofErr: nil}, nil
+	}
+	// Snap state is chunked, generate edge proofs for verification.
+	tr, err := trie.New(root, dl.triedb)
+	if err != nil {
+		log.Error("Missing trie", "root", root, "err", err)
+		return nil, err
+	}
+	// Firstly find out the key of last iterated element.
+	var last []byte
+	if len(keys) > 0 {
+		last = keys[len(keys)-1]
+	}
+	// Generate the Merkle proofs for the first and last element
+	if origin == nil {
+		origin = common.Hash{}.Bytes()
+	}
+	if err := tr.Prove(origin, 0, proof); err != nil {
+		log.Debug("Failed to prove range", "kind", kind, "origin", origin, "err", err)
+		return &proofResult{keys: keys, vals: vals, diskMore: diskMore, trieMore: false, proofErr: err, tr: tr}, nil
+	}
+	if last != nil {
+		if err := tr.Prove(last, 0, proof); err != nil {
+			log.Debug("Failed to prove range", "kind", kind, "last", last, "err", err)
+			return &proofResult{keys: keys, vals: vals, diskMore: diskMore, trieMore: false, proofErr: err, tr: tr}, nil
+		}
+	}
+	// Verify the state segment with range prover, ensure that all flat states
+	// in this range correspond to merkle trie.
+	_, _, _, cont, err := trie.VerifyRangeProof(root, origin, last, keys, vals, proof)
+	return &proofResult{keys: keys, vals: vals, diskMore: diskMore, trieMore: cont, proofErr: err, tr: tr}, nil
+}
+
+// genRange generates the state segment with particular prefix. Generation can
+// either verify the correctness of existing state through rangeproof and skip
+// generation, or iterate trie to regenerate state on demand.
+func (dl *diskLayer) genRange(root common.Hash, prefix []byte, kind string, origin []byte, max int, stats *generatorStats, onState func(key []byte, val []byte, write bool, delete bool) error, valueConvertFn func([]byte) ([]byte, error)) (bool, []byte, error) {
```

Comment:
Let's define a type for these callbacks :D Looking at this header makes my head spin